File: javascript/luhn/luhn.js
const add = (accumulator, currentValue) => accumulator + currentValue;

export class Luhn {
  constructor(input) {
    this.numbers = input
      .replace(/\s/g, "")
      .split("")
      .reverse()
      .map(char => parseInt(char));
    this.multiplyEvens();
    this.subtractNine();
  }

  get valid() {
    if (this.numbers.length < 2)
      return false;

    return this.numbers.reduce(add) % 10 == 0;
  }

  multiplyEvens() {
    this.numbers = this.numbers.map(function(number, i) {
      return i % 2 !== 0 ? number * 2 : number;
    });
  }

  subtractNine() {
    this.numbers = this.numbers.map(function(number) {
      return number > 9 ? number - 9 : number;
    });
  }
}

File: ruby/luhn/luhn.rb
class Luhn
  def self.valid?(input)
    input.delete! ' '
    return false if input.match?(/^.{1}$|\D/)
    luhn = new(input)
    (luhn.sum % 10).zero?
  end

  def initialize(number_input)
    @numbers = number_input.reverse.each_char.map(&:to_i)
    multiply_evens
    subtract_nine
  end

  def multiply_evens
    @numbers.map!.with_index do |n, i|
      i.odd? ? n * 2 : n
    end
  end

  def subtract_nine
    @numbers.map! do |n|
      n > 9 ? n - 9 : n
    end
  end

  def sum
    @numbers.sum
  end
end
